Refactor Activator.java to improve database connection handling

marwyg opened the pull request:
Fixes #5952

### Description

Implemented a wait mechanism to ensure database availability before starting the application. This resolves issues where the application fails to start if the database is not yet available.

### To test this:

1. Make sure the database is not running
2. Start opencast
3. Observe the log, you should see, that Opencast retries to connect to the database
4. Start the Database
5. Observe the log, Opencast should continue the startup process without errors

### Test Scenario 2:

I also included a check for credentials, to test this, do the following:

1. Configure a wrong username or password in the `custom.properties` file
2. Make sure the database is not running
3. Start opencast
4. Observe the log: The startup of Opencast should fail with clear logging

marwyg committed:
Refactor Activator.java to improve database connection handling

Implemented a wait mechanism to ensure database availability before starting the application. This resolves issues where the application fails to start if the database is not yet available.

Fixes #5952: Opencast crashes on startup when database is not running.

lkiesow commented on modules/db/src/main/java/org/opencastproject/db/Activator.java:
```diff
@@ -198,6 +191,50 @@ public void start(BundleContext bundleContext) throws Exception {
+    logger.info("Testing connectivity to database at {}", jdbcUrl);
+
+    // Load the JDBC driver for our basic connection test
+    try {
+      Class.forName(jdbcDriver);
+    } catch (ClassNotFoundException e) {
+      logger.error("Couldn't load JDBC Driver {} for initial connection test", jdbcDriver);
+      throw new RuntimeException(e);
```
This is different from the connection settings we use above, Wouldn't it be better to use the same? In fact, couldn't you just put a loop around the `try-catch` block you removed above and modify the ctach block?

marwyg replied:
Thank you for the review! Creating a loop around the original connection method was my first approach. However, I chose a different method for connecting to the database because the `pooledDataSource` logs exceptions upon failure. This alternative approach helps maintain a cleaner log. But we can use the `pooledDataSource` if the exceptions are acceptable.

lkiesow commented on modules/db/src/main/java/org/opencastproject/db/Activator.java:
```diff
@@ -198,6 +191,50 @@ public void start(BundleContext bundleContext) throws Exception {
+          connectionEstablished = connection != null && !connection.isClosed();
+        } catch (SQLInvalidAuthorizationSpecException e) {
+          // Invalid credentials
+          String errorMsg = "Couldn't connect to database. Probably invalid database credentials. Check you config.";
+          logger.error(errorMsg);
+          throw new RuntimeException(errorMsg, e);
```
It may seem minor, but it's actually pretty annoying to log errors and then throw them. The thrown error will either be caught and handled properly, turning the error into a non-error, or it will stay an error and be logged again, causing the error to appear at two separate places in the logs.

So, just throw the error and do not log:
```suggestion
          throw new RuntimeException("Couldn't connect to database. Probably invalid database credentials. Check you config.";, e);
```

marwyg replied:
I removed it.

lkiesow reviewed (changes requested):
Overall, this looks good. I'm just not entirely shure about the connection method

marwyg committed:
Fix redundant logging

Remove redundant logging in the activator class, directly before throwing a RuntimeException with the same message.

Arnei reviewed (approved):
Works with my local setup and seems reasonable to me. Agree with Lars that we optimally would use the exact same connection settings, but since this only wants to be a simple a connection check, this is fine I think?

gregorydlogan reviewed (approved):
Works as expected, although I agree with Lars that it'd be better to log more and reuse the sensitive connection logic.

This retry code really shouldn't be needed most of the time, so if the exceptions are the issue then the adopter should probably fix their init system.

KatrinIhler commented:
FYI: Branch cut for OC 17 is on November 6!

@marwyg  @lkiesow Can you come to an agreement here on the last outstanding comment?

geichelberger reviewed (changes requested):
Wrapping an exception should add a value by throwing a more specific exception. The RuntimeException does not add this value, and because of the nested try-catch, the original exception is nested two times in a RuntimeException.

gregorydlogan commented:
Is this going to be moving forward in time for OC18?
